include/boost/math/statistics/detail/single_pass.hpp

```diff
@@ -106,90 +106,28 @@ ReturnType first_four_moments_sequential_impl(ForwardIterator first, ForwardIter
-    auto future_a {std::async(std::launch::async | std::launch::deferred, [first, range_a]() -> ReturnType
-    {
-        if(thread_counter + 2 <= num_threads && range_a > 10)
-        {
-            return first_four_moments_parallel_impl<ReturnType>(first, std::next(first, range_a));
-        }
-        else
-        {
-            return first_four_moments_sequential_impl<ReturnType>(first, std::next(first, range_a));
-        }
-    })};
-    auto future_b {std::async(std::launch::async | std::launch::deferred, [first, last, range_a]() -> ReturnType
-    {
-        if(thread_counter + 2 <= num_threads && range_a > 10)
-        {
-            return first_four_moments_parallel_impl<ReturnType>(std::next(first, range_a), last);
-        }
-        else
-        {
-            return first_four_moments_sequential_impl<ReturnType>(std::next(first, range_a), last);
-        }
-    })};
-
-    const auto results_a {future_a.get()};
-    const auto results_b {future_b.get()};
-    thread_counter.fetch_sub(2);
-
-    const Real M1_a = std::get<0>(results_a);
-    const Real M2_a = std::get<1>(results_a);
-    const Real M3_a = std::get<2>(results_a);
-    const Real M4_a = std::get<3>(results_a);
-
-    const Real M1_b = std::get<0>(results_b);
-    const Real M2_b = std::get<1>(results_b);
-    const Real M3_b = std::get<2>(results_b);
-    const Real M4_b = std::get<3>(results_b);
-
-    const Real n_ab = elements;
-    const Real delta = M1_b - M1_a;
-    
-    const Real M1_ab = (range_a * M1_a + range_b * M1_b) / n_ab;
-    const Real M2_ab = M2_a + M2_b + delta * delta * (range_a * range_b / n_ab);
-    const Real M3_ab = M3_a + M3_b + (delta * delta * delta) * range_a * range_b * (range_a - range_b) / (n_ab * n_ab)    
-                       + Real(3) * delta * (range_a * M2_b - range_b * M2_a) / n_ab;
-    const Real M4_ab = M4_a + M4_b + (delta * delta * delta * delta) * range_a * range_b * (range_a * range_a - range_a * range_b + range_b * range_b) / (n_ab * n_ab * n_ab)
-                       + Real(6) * delta * delta * (range_a * range_a * M2_b + range_b * range_b * M2_a) / (n_ab * n_ab) 
-                       + Real(4) * delta * (range_a * M3_b - range_b * M3_a) / n_ab;
-
-    return std::make_tuple(M1_ab, M2_ab, M3_ab, M4_ab, n_ab);
-}
-*/
-
+// https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Higher-order_statistics
+// EQN 3.1: https://www.osti.gov/servlets/purl/1426900
 template<typename ReturnType, typename ForwardIterator>
 ReturnType first_four_moments_parallel_impl(ForwardIterator first, ForwardIterator last)
 {
     using Real = typename std::tuple_element<0, ReturnType>::type;
 
     const auto elements = std::distance(first, last);
     static unsigned num_threads;
-    if(elements/10 >= std::thread::hardware_concurrency())
+    
+    // https://lemire.me/blog/2020/01/30/cost-of-a-thread-in-c-under-linux/
+    if(elements/2000 >= std::thread::hardware_concurrency())
     {
```

Comment:
What happens if `std::thread::hardware_concurrency()` returns 0?

Also, might be better to do `elements >= 2000*std::thread::hardware_concurrency()` than do the division.

Reply:
Am I correct in thinking that this model assumes that each thread generation costs 10us, sequentially, for 10j us of cost? I believe it costs the same to launch 10 threads as it does to launch 1.

Reply:
Ok, assuming it take 10µs to launch j threads, independent of j, the parallel calculation is faster whenever

10 + 5.13e-3 N/j <= 5.13e-3N => N >= 10^4j/5.13(j-1).

The j = 0 and j=1 case should shunt through the sequential case, obviously.

Reply:
```
diff --git a/include/boost/math/statistics/detail/single_pass.hpp b/include/boost/math/statistics/detail/single_pass.hpp
index 00b2d9e67..8559d0b31 100644
--- a/include/boost/math/statistics/detail/single_pass.hpp
+++ b/include/boost/math/statistics/detail/single_pass.hpp
@@ -114,20 +114,32 @@ ReturnType first_four_moments_parallel_impl(ForwardIterator first, ForwardIterat
     using Real = typename std::tuple_element<0, ReturnType>::type;
 
     const auto elements = std::distance(first, last);
-    static unsigned num_threads;
+    const unsigned max_concurrency = std::thread::hardware_concurrency() == 0 ? 2u : std::thread::hardware_concurrency();
+    unsigned num_threads = 2u;
     
+    // Threading is faster for: 10 + 5.13e-3 N/j <= 5.13e-3N => N >= 10^4j/5.13(j-1).
+    const auto parallel_lower_bound = 10e4*max_concurrency/(5.13*(max_concurrency-1));
+    const auto parallel_upper_bound = 10e4*2/5.13; // j = 2
+
     // https://lemire.me/blog/2020/01/30/cost-of-a-thread-in-c-under-linux/
-    if(elements/2000 >= std::thread::hardware_concurrency())
+    if(elements < parallel_lower_bound)
     {
-        num_threads = std::thread::hardware_concurrency();
+        return detail::first_four_moments_sequential_impl<ReturnType>(first, last);
     }
-    else if(elements < 4000)
+    else if(elements >= parallel_upper_bound)
     {
-        return detail::first_four_moments_sequential_impl<ReturnType>(first, last);
+        num_threads = max_concurrency;
     }
     else
     {
-        num_threads = elements/2000;
+        for(unsigned i = 3; i < max_concurrency; ++i)
+        {
+            if(parallel_lower_bound < 10e4*i/(5.13*(i-1)))
+            {
+                num_threads = i;
+                break;
+            }
+        }
```
Here is a quick diff. It should phase in the number of threads up to the maximum.